File: examples/python/custom_postprocessor.py
"""Custom PostProcessor Example.

Demonstrates implementing custom post-processor plugins.
"""

import re
from datetime import datetime

from kreuzberg import (
    ExtractionResult,
    clear_post_processors,
    extract_file_sync,
    register_post_processor,
    unregister_post_processor,
)


class MetadataEnricher:
    """Post-processor that enriches extraction results with metadata."""

    def name(self) -> str:
        return "metadata_enricher"

    def process(self, result: ExtractionResult) -> ExtractionResult:
        """Add statistical metadata to the result."""
        content = result.content

        result.metadata["processed_at"] = datetime.now().isoformat()
        result.metadata["word_count"] = len(content.split())
        result.metadata["char_count"] = len(content)
        result.metadata["line_count"] = len(content.splitlines())
        result.metadata["has_content"] = bool(content.strip())

        result.metadata["has_urls"] = bool(re.search(r"https?://", content))
        result.metadata["has_emails"] = bool(re.search(r"\S+@\S+\.\S+", content))
        result.metadata["has_phone_numbers"] = bool(re.search(r"\d{3}[-.]?\d{3}[-.]?\d{4}", content))

        return result


class PIIRedactor:
    """Post-processor that redacts Personally Identifiable Information."""

    def name(self) -> str:
        return "pii_redactor"

    def process(self, result: ExtractionResult) -> ExtractionResult:
        """Redact PII from content."""
        content = result.content

        content = re.sub(r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b", "[EMAIL REDACTED]", content)

        content = re.sub(r"\(\d{3}\)\s*\d{3}[-.]?\d{4}", "[PHONE REDACTED]", content)
        content = re.sub(r"\d{3}[-.]?\d{3}[-.]?\d{4}", "[PHONE REDACTED]", content)

        content = re.sub(r"\b\d{3}-\d{2}-\d{4}\b", "[SSN REDACTED]", content)

        content = re.sub(r"\b\d{4}[-\s]?\d{4}[-\s]?\d{4}[-\s]?\d{4}\b", "[CARD REDACTED]", content)

        result.content = content
        result.metadata["pii_redacted"] = True

        return result


class TextNormalizer:
    """Post-processor that normalizes text formatting."""

    def name(self) -> str:
        return "text_normalizer"

    def process(self, result: ExtractionResult) -> ExtractionResult:
        """Normalize text formatting."""
        content = result.content

        content = re.sub(r" +", " ", content)
        content = re.sub(r"\n{3,}", "\n\n", content)

        lines = [line for line in content.splitlines() if line.strip()]
        content = "\n".join(lines)

        content = content.strip()

        result.content = content
        result.metadata["text_normalized"] = True

        return result


class LanguageTranslator:
    """Post-processor that translates content to target language.

    Note: This is a mock example. In production, use a translation API.
    """

    def __init__(self, target_language: str = "en") -> None:
        self.target_language = target_language

    def name(self) -> str:
        return "language_translator"

    def process(self, result: ExtractionResult) -> ExtractionResult:
        """Translate content to target language."""
        result.metadata["translated_to"] = self.target_language
        result.metadata["original_language"] = result.detected_languages[0] if result.detected_languages else "unknown"

        return result


class SummaryGenerator:
    """Post-processor that generates a summary of the content."""

    def __init__(self, max_summary_length: int = 500) -> None:
        self.max_summary_length = max_summary_length

    def name(self) -> str:
        return "summary_generator"

    def process(self, result: ExtractionResult) -> ExtractionResult:
        """Generate content summary."""
        content = result.content

        summary = content[: self.max_summary_length]

        if len(content) > self.max_summary_length:
            last_period = summary.rfind(".")
            last_newline = summary.rfind("\n")
            break_point = max(last_period, last_newline)

            if break_point > 0:
                summary = summary[: break_point + 1]
            else:
                summary += "..."

        result.metadata["summary"] = summary.strip()
        result.metadata["is_truncated"] = len(content) > self.max_summary_length

        return result


class KeywordExtractor:
    """Post-processor that extracts keywords from content."""

    def name(self) -> str:
        return "keyword_extractor"

    def process(self, result: ExtractionResult) -> ExtractionResult:
        """Extract keywords from content."""
        content = result.content.lower()

        stopwords = {
            "the",
            "a",
            "an",
            "and",
            "or",
            "but",
            "in",
            "on",
            "at",
            "to",
            "for",
            "of",
            "with",
            "is",
            "was",
            "are",
            "were",
            "be",
            "been",
            "being",
        }

        words = re.findall(r"\b[a-z]{4,}\b", content)

        word_freq: dict[str, int] = {}
        for word in words:
            if word not in stopwords:
                word_freq[word] = word_freq.get(word, 0) + 1

        keywords = sorted(word_freq.items(), key=lambda x: x[1], reverse=True)[:10]
        result.metadata["keywords"] = [word for word, _ in keywords]

        return result


def main() -> None:
    register_post_processor(MetadataEnricher())
    register_post_processor(PIIRedactor())
    register_post_processor(TextNormalizer())
    register_post_processor(LanguageTranslator(target_language="en"))
    register_post_processor(SummaryGenerator(max_summary_length=300))
    register_post_processor(KeywordExtractor())

    extract_file_sync("document.pdf")

    unregister_post_processor("pii_redactor")

    extract_file_sync("document.pdf")

    clear_post_processors()

    extract_file_sync("document.pdf")

    register_post_processor(MetadataEnricher())
    register_post_processor(SummaryGenerator(max_summary_length=200))

    extract_file_sync("document.pdf")


if __name__ == "__main__":
    main()

File: examples/typescript/custom-postprocessor.ts
/**
 * Custom PostProcessor Example
 *
 * Demonstrates implementing custom post-processor plugins.
 */

import {
	clearPostProcessors,
	type ExtractionResult,
	extractFile,
	extractFileSync,
	type PostProcessorProtocol,
	registerPostProcessor,
	unregisterPostProcessor,
} from "@goldziher/kreuzberg";

/**
 * Post-processor that enriches extraction results with metadata.
 */
class MetadataEnricher implements PostProcessorProtocol {
	name(): string {
		return "metadata_enricher";
	}

	process(result: ExtractionResult): ExtractionResult {
		const content = result.content;

		result.metadata.processed_at = new Date().toISOString();
		result.metadata.word_count = content.split(/\s+/).filter((w) => w.length > 0).length;
		result.metadata.char_count = content.length;
		result.metadata.line_count = content.split("\n").length;
		result.metadata.has_content = content.trim().length > 0;

		result.metadata.has_urls = /https?:\/\//.test(content);
		result.metadata.has_emails = /\S+@\S+\.\S+/.test(content);
		result.metadata.has_phone_numbers = /\d{3}[-.]?\d{3}[-.]?\d{4}/.test(content);

		console.log(`[MetadataEnricher] Added statistics: ${result.metadata.word_count} words`);

		return result;
	}
}

/**
 * Post-processor that redacts Personally Identifiable Information.
 */
class PIIRedactor implements PostProcessorProtocol {
	name(): string {
		return "pii_redactor";
	}

	process(result: ExtractionResult): ExtractionResult {
		let content = result.content;

		content = content.replace(/\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/g, "[EMAIL REDACTED]");

		content = content.replace(/\(\d{3}\)\s*\d{3}[-.]?\d{4}/g, "[PHONE REDACTED]");
		content = content.replace(/\d{3}[-.]?\d{3}[-.]?\d{4}/g, "[PHONE REDACTED]");

		content = content.replace(/\b\d{3}-\d{2}-\d{4}\b/g, "[SSN REDACTED]");

		content = content.replace(/\b\d{4}[-\s]?\d{4}[-\s]?\d{4}[-\s]?\d{4}\b/g, "[CARD REDACTED]");

		result.content = content;
		result.metadata.pii_redacted = true;

		console.log("[PIIRedactor] Redacted PII from content");

		return result;
	}
}

/**
 * Post-processor that normalizes text formatting.
 */
class TextNormalizer implements PostProcessorProtocol {
	name(): string {
		return "text_normalizer";
	}

	process(result: ExtractionResult): ExtractionResult {
		let content = result.content;

		content = content.replace(/ +/g, " ");
		content = content.replace(/\n{3,}/g, "\n\n");

		const lines = content.split("\n").filter((line) => line.trim().length > 0);
		content = lines.join("\n");

		content = content.normalize("NFC").trim();

		result.content = content;
		result.metadata.text_normalized = true;

		console.log("[TextNormalizer] Normalized text formatting");

		return result;
	}
}

/**
 * Post-processor that generates a summary of the content.
 */
class SummaryGenerator implements PostProcessorProtocol {
	constructor(private maxSummaryLength: number = 500) {}

	name(): string {
		return "summary_generator";
	}

	process(result: ExtractionResult): ExtractionResult {
		const content = result.content;
		let summary = content.substring(0, this.maxSummaryLength);

		if (content.length > this.maxSummaryLength) {
			const lastPeriod = summary.lastIndexOf(".");
			const lastNewline = summary.lastIndexOf("\n");
			const breakPoint = Math.max(lastPeriod, lastNewline);

			if (breakPoint > 0) {
				summary = summary.substring(0, breakPoint + 1);
			} else {
				summary += "...";
			}
		}

		result.metadata.summary = summary.trim();
		result.metadata.is_truncated = content.length > this.maxSummaryLength;

		console.log(`[SummaryGenerator] Generated summary: ${summary.length} chars`);

		return result;
	}
}

/**
 * Post-processor that extracts keywords from content.
 */
class KeywordExtractor implements PostProcessorProtocol {
	name(): string {
		return "keyword_extractor";
	}

	process(result: ExtractionResult): ExtractionResult {
		const content = result.content.toLowerCase();

		const stopwords = new Set([
			"the",
			"a",
			"an",
			"and",
			"or",
			"but",
			"in",
			"on",
			"at",
			"to",
			"for",
			"of",
			"with",
			"is",
			"was",
			"are",
			"were",
			"be",
			"been",
			"being",
		]);

		const words = content.match(/\b[a-z]{4,}\b/g) || [];

		const wordFreq = new Map<string, number>();
		for (const word of words) {
			if (!stopwords.has(word)) {
				wordFreq.set(word, (wordFreq.get(word) || 0) + 1);
			}
		}

		const keywords = Array.from(wordFreq.entries())
			.sort((a, b) => b[1] - a[1])
			.slice(0, 10)
			.map(([word]) => word);

		result.metadata.keywords = keywords;

		console.log(`[KeywordExtractor] Extracted ${keywords.length} keywords`);

		return result;
	}
}

/**
 * Async post-processor that calls external API for enrichment.
 */
class ExternalAPIEnricher implements PostProcessorProtocol {
	constructor(
		private apiUrl: string,
		private apiKey: string,
	) {}

	name(): string {
		return "external_api_enricher";
	}

	async process(result: ExtractionResult): Promise<ExtractionResult> {
		try {
			console.log(`[ExternalAPIEnricher] Calling API: ${this.apiUrl}`);

			result.metadata.external_data = {
				sentiment: "positive",
				topics: ["technology", "business"],
				language_confidence: 0.95,
			};

			console.log("[ExternalAPIEnricher] API enrichment complete");
		} catch (error) {
			console.error("[ExternalAPIEnricher] API call failed:", error);
			result.metadata.api_error = error instanceof Error ? error.message : "Unknown error";
		}

		return result;
	}
}

async function main() {
	console.log("=== Registering Post-Processors ===");
	registerPostProcessor(new MetadataEnricher());
	registerPostProcessor(new PIIRedactor());
	registerPostProcessor(new TextNormalizer());
	registerPostProcessor(new SummaryGenerator(300));
	registerPostProcessor(new KeywordExtractor());
	registerPostProcessor(new ExternalAPIEnricher("https://api.example.com", "api-key"));

	console.log("Registered 6 post-processors\n");

	console.log("=== Extraction with Post-Processors ===");
	const result = await extractFile("document.pdf");

	console.log("\nFinal result:");
	console.log(`  Content length: ${result.content.length} chars`);
	console.log(`  Word count: ${result.metadata.word_count}`);
	console.log(`  PII redacted: ${result.metadata.pii_redacted}`);
	console.log(`  Text normalized: ${result.metadata.text_normalized}`);
	console.log(`  Summary length: ${(result.metadata.summary as string)?.length || 0}`);
	console.log(`  Keywords: ${(result.metadata.keywords as string[])?.slice(0, 5)}`);

	console.log("\n=== Unregister Post-Processor ===");
	unregisterPostProcessor("pii_redactor");
	console.log("Unregistered: pii_redactor");

	const result2 = extractFileSync("document.pdf");
	console.log(`PII redacted: ${result2.metadata.pii_redacted || false}`);

	console.log("\n=== Clear All Post-Processors ===");
	clearPostProcessors();
	console.log("Cleared all post-processors");

	const result3 = extractFileSync("document.pdf");
	console.log(`Word count (should be missing): ${result3.metadata.word_count}`);

	console.log("\n=== Selective Post-Processing ===");
	registerPostProcessor(new MetadataEnricher());
	registerPostProcessor(new SummaryGenerator(200));

	const result4 = extractFileSync("document.pdf");
	console.log(`Only metadata enrichment and summary: ${Object.keys(result4.metadata)}`);
}

main().catch(console.error);
